mc2/controllers/docker/tests/test_docker_controller.py

```diff
@@ -2,10 +2,162 @@
 import responses
 from django.conf import settings
 from django.contrib.auth.models import User
+from hypothesis import given
+from hypothesis.extra.django import TestCase
+from hypothesis.strategies import text, random_module, lists, just
+
+from mc2.controllers.docker.tests.hypothesis_helper import (
+    models, DEFAULT_VALUE)
+
+from mc2.controllers.base.models import EnvVariable, MarathonLabel
 from mc2.controllers.base.tests.base import ControllerBaseTestCase
 from mc2.controllers.docker.models import DockerController
 
 
+def add_envvars(controller):
+    envvars = lists(models(EnvVariable, controller=just(controller)))
+    return envvars.map(lambda _: controller)
+
+
+def add_labels(controller):
+    labels = lists(models(MarathonLabel, controller=just(controller)))
+    return labels.map(lambda _: controller)
+
+
+def docker_controller(with_envvars=True, with_labels=True, **kw):
+    """
+    Strategy to generate a controller model with (optional) envvars and labels.
+    """
+    # The slug is used in places where whitespace and colos are problematic, so
+    # we remove them from the generated value.
+    # TODO: Build a proper SlugField strategy.
+    # TODO: Figure out why the field validation isn't being applied.
+    kw.setdefault("slug", text().map(
+        lambda t: "".join(t.replace(":", "").split())))
+    kw.setdefault("owner", models(User))
+    # The model generator sees `controller_ptr` (from the PolymorphicModel
+    # magic) as a mandatory field and objects if we don't provide a value for
+    # it.
+    controller = models(DockerController, controller_ptr=DEFAULT_VALUE, **kw)
+    if with_envvars:
+        controller = controller.flatmap(add_envvars)
```

Comment:
The name comes from the functional programming world. The Hypothesis docs explain it better than I could: https://hypothesis.readthedocs.org/en/release/data.html#chaining-strategies-together